src/v/redpanda/memory_sampling.cc

```diff
@@ -0,0 +1,72 @@
+#include "redpanda/memory_sampling.h"
+
+#include "ssx/future-util.h"
+#include "vlog.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/memory.hh>
+#include <seastar/core/sstring.hh>
+#include <seastar/util/memory_diagnostics.hh>
+
+#include <fmt/core.h>
+
+constexpr std::string_view diagnostics_header() {
+    return "Top-N alloc sites - size count stack:\n";
+}
+
+constexpr std::string_view allocation_site_format_str() { return "{} {} {}\n"; }
+
+/// Put `top_n` allocation sites into the front of `allocation_sites`
+void top_n_allocation_sites(
+  std::vector<ss::memory::allocation_site>& allocation_sites, size_t top_n) {
+    std::partial_sort(
+      allocation_sites.begin(),
+      allocation_sites.begin() + top_n,
+      allocation_sites.end(),
+      [](const auto& lhs, const auto& rhs) { return lhs.size > rhs.size; });
+}
+
+ss::sstring memory_sampling::format_allocation_site(
+  const ss::memory::allocation_site& alloc_site) {
+    return fmt::format(
```

Comment:
This could be impreved by implementing the formatter. Also, checkout out `ssx::sformat`, which will directly render an `ss::sstring` instead of copying it from a `std::string`